<!DOCTYPE html>

<html class="no-js" data-content_root="./" lang="en">
<head><meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="light dark" name="color-scheme"/><meta content="width=device-width, initial-scale=1" name="viewport">
<script async="" src="https://www.googletagmanager.com/gtag/js?id="></script>
<script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            
            gtag('config', '');
            
        </script>
<meta content="How to prepare PoseBench data" property="og:title">
<meta content="website" property="og:type">
<meta content="data_preparation.html" property="og:url"/>
<meta content="PoseBench" property="og:site_name"/>
<meta content="Downloading Astex, PoseBusters, DockGen, and CASP15 data: Downloading benchmark method predictions: Downloading benchmark method interactions: Downloading sequence databases (required only for Rose..." property="og:description"/>
<meta content="1146" property="og:image:width"/>
<meta content="600" property="og:image:height"/>
<meta content="_images/social_previews/summary_data_preparation_bc9d7f9e.png" property="og:image"/>
<meta content="Downloading Astex, PoseBusters, DockGen, and CASP15 data: Downloading benchmark method predictions: Downloading benchmark method interactions: Downloading sequence databases (required only for Rose..." property="og:image:alt"/>
<meta content="Downloading Astex, PoseBusters, DockGen, and CASP15 data: Downloading benchmark method predictions: Downloading benchmark method interactions: Downloading sequence databases (required only for Rose..." name="description"/>
<meta content="summary_large_image" name="twitter:card"/>
<link href="genindex.html" rel="index" title="Index"/><link href="search.html" rel="search" title="Search"/><link href="available_methods.html" rel="next" title="Available inference methods"/><link href="tutorials.html" rel="prev" title="Tutorials"/>
<link as="image" href="_static/WorkBench.jpeg" rel="prefetch"/>
<!-- Generated with Sphinx 8.1.3 and Furo 2025.09.25 -->
<title>How to prepare PoseBench data - PoseBench 1.0.0</title>
<link href="_static/pygments.css?v=d111a655" rel="stylesheet" type="text/css">
<link href="_static/styles/furo.css?v=580074bf" rel="stylesheet" type="text/css">
<link href="_static/copybutton.css?v=76b2166b" rel="stylesheet" type="text/css">
<link href="_static/tabs.css?v=4c969af8" rel="stylesheet" type="text/css">
<link href="_static/sphinx-codeautolink.css?v=b2176991" rel="stylesheet" type="text/css">
<link href="_static/styles/furo-extensions.css?v=8dab3a3b" rel="stylesheet" type="text/css"/>
<style>
  body {
    --color-code-background: #f2f2f2;
  --color-code-foreground: #1e1e1e;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></link></link></link></link></link></meta></meta></meta></head>
<body>
<script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
<svg style="display: none;" xmlns="http://www.w3.org/2000/svg">
<symbol id="svg-toc" viewbox="0 0 24 24">
<title>Contents</title>
<svg fill="currentColor" stroke="currentColor" stroke-width="0" viewbox="0 0 1024 1024">
<path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"></path>
</svg>
</symbol>
<symbol id="svg-menu" viewbox="0 0 24 24">
<title>Menu</title>
<svg class="feather-menu" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<line x1="3" x2="21" y1="12" y2="12"></line>
<line x1="3" x2="21" y1="6" y2="6"></line>
<line x1="3" x2="21" y1="18" y2="18"></line>
</svg>
</symbol>
<symbol id="svg-arrow-right" viewbox="0 0 24 24">
<title>Expand</title>
<svg class="feather-chevron-right" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<polyline points="9 18 15 12 9 6"></polyline>
</svg>
</symbol>
<symbol id="svg-sun" viewbox="0 0 24 24">
<title>Light mode</title>
<svg class="feather-sun" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<circle cx="12" cy="12" r="5"></circle>
<line x1="12" x2="12" y1="1" y2="3"></line>
<line x1="12" x2="12" y1="21" y2="23"></line>
<line x1="4.22" x2="5.64" y1="4.22" y2="5.64"></line>
<line x1="18.36" x2="19.78" y1="18.36" y2="19.78"></line>
<line x1="1" x2="3" y1="12" y2="12"></line>
<line x1="21" x2="23" y1="12" y2="12"></line>
<line x1="4.22" x2="5.64" y1="19.78" y2="18.36"></line>
<line x1="18.36" x2="19.78" y1="5.64" y2="4.22"></line>
</svg>
</symbol>
<symbol id="svg-moon" viewbox="0 0 24 24">
<title>Dark mode</title>
<svg class="icon-tabler-moon" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M0 0h24v24H0z" fill="none" stroke="none"></path>
<path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z"></path>
</svg>
</symbol>
<symbol id="svg-sun-with-moon" viewbox="0 0 24 24">
<title>Auto light/dark, in light mode</title>
<svg class="icon-custom-derived-from-feather-sun-and-tabler-moon" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z" style="opacity: 50%"></path>
<line x1="14.5" x2="14.5" y1="3.25" y2="1.25"></line>
<line x1="14.5" x2="14.5" y1="15.85" y2="17.85"></line>
<line x1="10.044" x2="8.63" y1="5.094" y2="3.68"></line>
<line x1="19" x2="20.414" y1="14.05" y2="15.464"></line>
<line x1="8.2" x2="6.2" y1="9.55" y2="9.55"></line>
<line x1="20.8" x2="22.8" y1="9.55" y2="9.55"></line>
<line x1="10.044" x2="8.63" y1="14.006" y2="15.42"></line>
<line x1="19" x2="20.414" y1="5.05" y2="3.636"></line>
<circle cx="14.5" cy="9.55" r="3.6"></circle>
</svg>
</symbol>
<symbol id="svg-moon-with-sun" viewbox="0 0 24 24">
<title>Auto light/dark, in dark mode</title>
<svg class="icon-custom-derived-from-feather-sun-and-tabler-moon" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"></path>
<line style="opacity: 50%" x1="18" x2="18" y1="3.705" y2="2.5"></line>
<line style="opacity: 50%" x1="18" x2="18" y1="11.295" y2="12.5"></line>
<line style="opacity: 50%" x1="15.316" x2="14.464" y1="4.816" y2="3.964"></line>
<line style="opacity: 50%" x1="20.711" x2="21.563" y1="10.212" y2="11.063"></line>
<line style="opacity: 50%" x1="14.205" x2="13.001" y1="7.5" y2="7.5"></line>
<line style="opacity: 50%" x1="21.795" x2="23" y1="7.5" y2="7.5"></line>
<line style="opacity: 50%" x1="15.316" x2="14.464" y1="10.184" y2="11.036"></line>
<line style="opacity: 50%" x1="20.711" x2="21.563" y1="4.789" y2="3.937"></line>
<circle cx="18" cy="7.5" r="2.169" style="opacity: 50%"></circle>
</svg>
</symbol>
<symbol id="svg-pencil" viewbox="0 0 24 24">
<svg class="icon-tabler-pencil-code" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"></path>
<path d="M13.5 6.5l4 4"></path>
<path d="M20 21l2 -2l-2 -2"></path>
<path d="M17 17l-2 2l2 2"></path>
</svg>
</symbol>
<symbol id="svg-eye" viewbox="0 0 24 24">
<svg class="icon-tabler-eye-code" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1" viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M0 0h24v24H0z" fill="none" stroke="none"></path>
<path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0"></path>
<path d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008"></path>
<path d="M20 21l2 -2l-2 -2"></path>
<path d="M17 17l-2 2l2 2"></path>
</svg>
</symbol>
</svg>
<input aria-label="Toggle site navigation sidebar" class="sidebar-toggle" id="__navigation" name="__navigation" type="checkbox"/>
<input aria-label="Toggle table of contents sidebar" class="sidebar-toggle" id="__toc" name="__toc" type="checkbox"/>
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>
<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>
<div class="page">
<header class="mobile-header">
<div class="header-left">
<label class="nav-overlay-icon" for="__navigation">
<span class="icon"><svg><use href="#svg-menu"></use></svg></span>
</label>
</div>
<div class="header-center">
<a href="index.html"><div class="brand">PoseBench 1.0.0</div></a>
</div>
<div class="header-right">
<div class="theme-toggle-container theme-toggle-header">
<button aria-label="Toggle Light / Dark / Auto color theme" class="theme-toggle">
<svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
<svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
<svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
<svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
</button>
</div>
<label class="toc-overlay-icon toc-header-icon" for="__toc">
<span class="icon"><svg><use href="#svg-toc"></use></svg></span>
</label>
</div>
</header>
<aside class="sidebar-drawer">
<div class="sidebar-container">
<div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
<div class="sidebar-logo-container">
<img alt="Logo" class="sidebar-logo" src="_static/WorkBench.jpeg"/>
</div>
<span class="sidebar-brand-text">PoseBench 1.0.0</span>
</a><form action="search.html" class="sidebar-search-container" method="get" role="search">
<input aria-label="Search" class="sidebar-search" name="q" placeholder="Search"/>
<input name="check_keywords" type="hidden" value="yes"/>
<input name="area" type="hidden" value="default"/>
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">How to prepare <cite>PoseBench</cite> data</a></li>
<li class="toctree-l1"><a class="reference internal" href="available_methods.html">Available inference methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="sweep_inference.html">How to run a sweep of benchmarking experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="method_inference.html">How to run inference with individual methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="ensemble_inference.html">How to run inference with a method ensemble</a></li>
<li class="toctree-l1"><a class="reference internal" href="comparative_plots.html">How to create comparative plots of inference results</a></li>
<li class="toctree-l1"><a class="reference internal" href="for_developers.html">For developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing_this_work.html">Citing this work</a></li>
<li class="toctree-l1"><a class="reference internal" href="bonus.html">Bonus</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Default Configs</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="configs/analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="configs/data.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="configs/model.html">Model</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.binding_site_crop_preparation.html">Binding site crop preparation</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.complex_alignment.html">Complex alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.inference_relaxation.html">Inference relaxation</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.minimize_energy.html">Minimize energy</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.ensemble_generation.html">Ensemble generation</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.data_utils.html">Data utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.model_utils.html">Model utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.utils.html">General utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules/posebench.resolvers.html">OmegaConf resolvers</a></li>
</ul>
</div>
</div>
</div>
</div>
</aside>
<div class="main">
<div class="content">
<div class="article-container">
<a class="back-to-top muted-link" href="#">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
</svg>
<span>Back to top</span>
</a>
<div class="content-icon-container">
<div class="view-this-page">
<a class="muted-link" href="_sources/data_preparation.rst.txt" title="View this page">
<svg><use href="#svg-eye"></use></svg>
<span class="visually-hidden">View this page</span>
</a>
</div>
<div class="theme-toggle-container theme-toggle-content">
<button aria-label="Toggle Light / Dark / Auto color theme" class="theme-toggle">
<svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
<svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
<svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
<svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
</button>
</div>
<label class="toc-overlay-icon toc-content-icon" for="__toc">
<span class="icon"><svg><use href="#svg-toc"></use></svg></span>
</label>
</div>
<article id="furo-main-content" role="main">
<section id="how-to-prepare-posebench-data">
<h1>How to prepare <cite>PoseBench</cite> data<a class="headerlink" href="#how-to-prepare-posebench-data" title="Link to this heading">¶</a></h1>
<section id="downloading-astex-posebusters-dockgen-and-casp15-data">
<h2>Downloading Astex, PoseBusters, DockGen, and CASP15 data<a class="headerlink" href="#downloading-astex-posebusters-dockgen-and-casp15-data" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># fetch, extract, and clean-up preprocessed Astex Diverse, PoseBusters Benchmark, DockGen, and CASP15 data (~3 GB) #</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/astex_diverse_set.tar.gz
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/posebusters_benchmark_set.tar.gz
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/dockgen_set.tar.gz
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/casp15_set.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>astex_diverse_set.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>posebusters_benchmark_set.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>dockgen_set.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>casp15_set.tar.gz
rm<span class="w"> </span>astex_diverse_set.tar.gz
rm<span class="w"> </span>posebusters_benchmark_set.tar.gz
rm<span class="w"> </span>dockgen_set.tar.gz
rm<span class="w"> </span>casp15_set.tar.gz
</pre></div>
</div>
</section>
<section id="downloading-benchmark-method-predictions">
<h2>Downloading benchmark method predictions<a class="headerlink" href="#downloading-benchmark-method-predictions" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># fetch, extract, and clean-up benchmark method predictions to reproduce paper results (~19 GB) #</span>
<span class="c1"># AutoDock Vina predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/vina_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>vina_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>vina_benchmark_method_predictions.tar.gz
<span class="c1"># DiffDock predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/diffdock_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>diffdock_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>diffdock_benchmark_method_predictions.tar.gz
<span class="c1"># DynamicBind predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/dynamicbind_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>dynamicbind_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>dynamicbind_benchmark_method_predictions.tar.gz
<span class="c1"># NeuralPLexer predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/neuralplexer_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>neuralplexer_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>neuralplexer_benchmark_method_predictions.tar.gz
<span class="c1"># RoseTTAFold-All-Atom predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/rfaa_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>rfaa_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>rfaa_benchmark_method_predictions.tar.gz
<span class="c1"># Chai-1 predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/chai_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>chai_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>chai_benchmark_method_predictions.tar.gz
<span class="c1"># Boltz-1 predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/boltz_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>boltz_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>boltz_benchmark_method_predictions.tar.gz
<span class="c1"># AlphaFold 3 predictions and results</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/af3_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>af3_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>af3_benchmark_method_predictions.tar.gz
<span class="c1"># CASP15 predictions and results for all methods</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/casp15_benchmark_method_predictions.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>casp15_benchmark_method_predictions.tar.gz
rm<span class="w"> </span>casp15_benchmark_method_predictions.tar.gz
</pre></div>
</div>
</section>
<section id="downloading-benchmark-method-interactions">
<h2>Downloading benchmark method interactions<a class="headerlink" href="#downloading-benchmark-method-interactions" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># fetch, extract, and clean-up benchmark method interactions to reproduce paper results (~12 GB) #</span>
<span class="c1"># cached ProLIF interactions for notebook plots</span>
wget<span class="w"> </span>https://zenodo.org/records/17536252/files/posebench_notebooks.tar.gz
tar<span class="w"> </span>-xzf<span class="w"> </span>posebench_notebooks.tar.gz
rm<span class="w"> </span>posebench_notebooks.tar.gz
</pre></div>
</div>
</section>
<section id="downloading-sequence-databases-required-only-for-rosettafold-all-atom-inference">
<h2>Downloading sequence databases (required only for RoseTTAFold-All-Atom inference)<a class="headerlink" href="#downloading-sequence-databases-required-only-for-rosettafold-all-atom-inference" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># acquire multiple sequence alignment databases for RoseTTAFold-All-Atom (~2.5 TB)</span>
<span class="nb">cd</span><span class="w"> </span>forks/RoseTTAFold-All-Atom/

<span class="c1"># uniref30 [46G]</span>
wget<span class="w"> </span>http://wwwuser.gwdg.de/~compbiol/uniclust/2020_06/UniRef30_2020_06_hhsuite.tar.gz
mkdir<span class="w"> </span>-p<span class="w"> </span>UniRef30_2020_06
tar<span class="w"> </span>xfz<span class="w"> </span>UniRef30_2020_06_hhsuite.tar.gz<span class="w"> </span>-C<span class="w"> </span>./UniRef30_2020_06

<span class="c1"># BFD [272G]</span>
wget<span class="w"> </span>https://bfd.mmseqs.com/bfd_metaclust_clu_complete_id30_c90_final_seq.sorted_opt.tar.gz
mkdir<span class="w"> </span>-p<span class="w"> </span>bfd
tar<span class="w"> </span>xfz<span class="w"> </span>bfd_metaclust_clu_complete_id30_c90_final_seq.sorted_opt.tar.gz<span class="w"> </span>-C<span class="w"> </span>./bfd

<span class="c1"># structure templates [81G] (including *_a3m.ffdata, *_a3m.ffindex)</span>
wget<span class="w"> </span>https://files.ipd.uw.edu/pub/RoseTTAFold/pdb100_2021Mar03.tar.gz
tar<span class="w"> </span>xfz<span class="w"> </span>pdb100_2021Mar03.tar.gz

<span class="nb">cd</span><span class="w"> </span>../../
</pre></div>
</div>
</section>
<section id="downloading-pdb-metadata">
<h2>Downloading PDB metadata<a class="headerlink" href="#downloading-pdb-metadata" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># download and extract the PDB's FASTA sequence files</span>
mkdir<span class="w"> </span>-p<span class="w"> </span>./data/pdb_data/
wget<span class="w"> </span>-P<span class="w"> </span>./data/pdb_data/<span class="w"> </span>https://files.rcsb.org/pub/pdb/derived_data/pdb_seqres.txt.gz
find<span class="w"> </span>./data/pdb_data/<span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span>-name<span class="w"> </span><span class="s2">"*.gz"</span><span class="w"> </span>-exec<span class="w"> </span>gzip<span class="w"> </span>-d<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
</pre></div>
</div>
</section>
<section id="predicting-apo-protein-structures-using-esmfold-optional-preprocessed-data-available">
<h2>Predicting apo protein structures using ESMFold (optional, preprocessed data available)<a class="headerlink" href="#predicting-apo-protein-structures-using-esmfold-optional-preprocessed-data-available" title="Link to this heading">¶</a></h2>
<p>First create all the corresponding FASTA files for each protein sequence</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>posebench/data/components/fasta_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>posebusters_benchmark
python3<span class="w"> </span>posebench/data/components/fasta_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>astex_diverse
python3<span class="w"> </span>posebench/data/components/fasta_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>dockgen
python3<span class="w"> </span>posebench/data/components/fasta_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>casp15
</pre></div>
</div>
<p>To generate the apo version of each protein structure,
create ESMFold-ready versions of the combined FASTA files
prepared above by the script <code class="docutils literal notranslate"><span class="pre">fasta_preparation.py</span></code>
for the PoseBusters Benchmark and Astex Diverse sets, respectively</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>posebench/data/components/esmfold_sequence_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>posebusters_benchmark
python3<span class="w"> </span>posebench/data/components/esmfold_sequence_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>astex_diverse
python3<span class="w"> </span>posebench/data/components/esmfold_sequence_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>dockgen
python3<span class="w"> </span>posebench/data/components/esmfold_sequence_preparation.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>casp15
</pre></div>
</div>
<p>Then, predict each apo protein structure using ESMFold’s batch
inference script</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>posebench/data/components/esmfold_batch_structure_prediction.py<span class="w"> </span>-i<span class="w"> </span>data/posebusters_benchmark_set/reference_posebusters_benchmark_esmfold_sequences.fasta<span class="w"> </span>-o<span class="w"> </span>data/posebusters_benchmark_set/posebusters_benchmark_esmfold_predicted_structures<span class="w"> </span>--skip-existing
python3<span class="w"> </span>posebench/data/components/esmfold_batch_structure_prediction.py<span class="w"> </span>-i<span class="w"> </span>data/astex_diverse_set/reference_astex_diverse_esmfold_sequences.fasta<span class="w"> </span>-o<span class="w"> </span>data/astex_diverse_set/astex_diverse_esmfold_predicted_structures<span class="w"> </span>--skip-existing
python3<span class="w"> </span>posebench/data/components/esmfold_batch_structure_prediction.py<span class="w"> </span>-i<span class="w"> </span>data/dockgen_set/reference_dockgen_esmfold_sequences.fasta<span class="w"> </span>-o<span class="w"> </span>data/dockgen_set/dockgen_esmfold_predicted_structures<span class="w"> </span>--skip-existing
python3<span class="w"> </span>posebench/data/components/esmfold_batch_structure_prediction.py<span class="w"> </span>-i<span class="w"> </span>data/casp15_set/reference_casp15_esmfold_sequences.fasta<span class="w"> </span>-o<span class="w"> </span>data/casp15_set/casp15_esmfold_predicted_structures<span class="w"> </span>--skip-existing
</pre></div>
</div>
<p><strong>NOTE:</strong> Having a CUDA-enabled device available when running ESMFold is highly recommended</p>
<p><strong>NOTE:</strong> ESMFold may not be able to predict apo protein structures for a handful of exceedingly-long (e.g., &gt;2000 token) input sequences</p>
<p>Lastly, align each apo protein structure to its corresponding
holo protein structure counterpart for each dataset, taking ligand
conformations into account during each alignment</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>activate<span class="w"> </span>PyMOL-PoseBench
python3<span class="w"> </span>posebench/data/components/protein_apo_to_holo_alignment.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>posebusters_benchmark<span class="w"> </span><span class="nv">processing_esmfold_structures</span><span class="o">=</span><span class="nb">true</span><span class="w"> </span><span class="nv">num_workers</span><span class="o">=</span><span class="m">1</span>
python3<span class="w"> </span>posebench/data/components/protein_apo_to_holo_alignment.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>astex_diverse<span class="w"> </span><span class="nv">processing_esmfold_structures</span><span class="o">=</span><span class="nb">true</span><span class="w"> </span><span class="nv">num_workers</span><span class="o">=</span><span class="m">1</span>
python3<span class="w"> </span>posebench/data/components/protein_apo_to_holo_alignment.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>dockgen<span class="w"> </span><span class="nv">processing_esmfold_structures</span><span class="o">=</span><span class="nb">true</span><span class="w"> </span><span class="nv">num_workers</span><span class="o">=</span><span class="m">1</span>
python3<span class="w"> </span>posebench/data/components/protein_apo_to_holo_alignment.py<span class="w"> </span><span class="nv">dataset</span><span class="o">=</span>casp15<span class="w"> </span><span class="nv">processing_esmfold_structures</span><span class="o">=</span><span class="nb">true</span><span class="w"> </span><span class="nv">num_workers</span><span class="o">=</span><span class="m">1</span>
conda<span class="w"> </span>deactivate
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The preprocessed Astex Diverse, PoseBusters Benchmark, DockGen, and CASP15 data (available via Zenodo at <a class="reference external" href="https://doi.org/10.5281/zenodo.17536252">https://doi.org/10.5281/zenodo.17536252</a>) provide pre-holo-aligned protein structures predicted by AlphaFold 3 (and alternatively MIT-licensed ESMFold) for these respective datasets. Accordingly, users must ensure their usage of such predicted protein structures from AlphaFold 3 aligns with AlphaFold 3’s Terms of Use at <a class="reference external" href="https://github.com/google-deepmind/alphafold3/blob/main/WEIGHTS_TERMS_OF_USE.md">https://github.com/google-deepmind/alphafold3/blob/main/WEIGHTS_TERMS_OF_USE.md</a>.</p>
</div>
</section>
</section>
</article>
</div>
<footer>
<div class="related-pages">
<a class="next-page" href="available_methods.html">
<div class="page-info">
<div class="context">
<span>Next</span>
</div>
<div class="title">Available inference methods</div>
</div>
<svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
</a>
<a class="prev-page" href="tutorials.html">
<svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
<div class="page-info">
<div class="context">
<span>Previous</span>
</div>
<div class="title">Tutorials</div>
</div>
</a>
</div>
<div class="bottom-of-page">
<div class="left-details">
<div class="copyright">
                Copyright © 2025, Alex Morehead
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
</div>
<div class="right-details">
</div>
</div>
</footer>
</div>
<aside class="toc-drawer">
<div class="toc-sticky toc-scroll">
<div class="toc-title-container">
<span class="toc-title">
            On this page
          </span>
</div>
<div class="toc-tree-container">
<div class="toc-tree">
<ul>
<li><a class="reference internal" href="#">How to prepare <cite>PoseBench</cite> data</a><ul>
<li><a class="reference internal" href="#downloading-astex-posebusters-dockgen-and-casp15-data">Downloading Astex, PoseBusters, DockGen, and CASP15 data</a></li>
<li><a class="reference internal" href="#downloading-benchmark-method-predictions">Downloading benchmark method predictions</a></li>
<li><a class="reference internal" href="#downloading-benchmark-method-interactions">Downloading benchmark method interactions</a></li>
<li><a class="reference internal" href="#downloading-sequence-databases-required-only-for-rosettafold-all-atom-inference">Downloading sequence databases (required only for RoseTTAFold-All-Atom inference)</a></li>
<li><a class="reference internal" href="#downloading-pdb-metadata">Downloading PDB metadata</a></li>
<li><a class="reference internal" href="#predicting-apo-protein-structures-using-esmfold-optional-preprocessed-data-available">Predicting apo protein structures using ESMFold (optional, preprocessed data available)</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>
</aside>
</div>
</div><script src="_static/jquery.js?v=5d32c60e"></script>
<script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
<script src="_static/documentation_options.js?v=8d563738"></script>
<script src="_static/doctools.js?v=9bcbadda"></script>
<script src="_static/sphinx_highlight.js?v=dc90522c"></script>
<script src="_static/scripts/furo.js?v=46bd48cc"></script>
<script src="_static/clipboard.min.js?v=a7894cd8"></script>
<script src="_static/copybutton.js?v=f281be69"></script>
<script src="_static/tabs.js?v=3ee01567"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
</body>
</html>