{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PoseBusters Benchmark Method Interaction Analysis Plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "import tempfile\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from beartype import beartype\n",
    "from beartype.typing import Any, Literal\n",
    "from Bio.PDB import PDBIO, PDBParser, Select\n",
    "from posecheck import PoseCheck\n",
    "from rdkit import Chem\n",
    "from tqdm import tqdm\n",
    "\n",
    "from posebench import resolve_method_ligand_dir, resolve_method_protein_dir\n",
    "from posebench.utils.data_utils import count_num_residues_in_pdb_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Configure packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.mode.copy_on_write = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General variables\n",
    "baseline_methods = [\n",
    "    \"diffdock\",\n",
    "    \"diffdockv1\",\n",
    "    \"fabind\",\n",
    "    \"dynamicbind\",\n",
    "    \"neuralplexer\",\n",
    "    \"rfaa\",\n",
    "    \"chai-lab\",\n",
    "    \"tulip\",\n",
    "    \"vina_diffdock\",\n",
    "    \"vina_p2rank\",\n",
    "    \"consensus_ensemble\",\n",
    "]\n",
    "max_num_repeats_per_method = (\n",
    "    1  # NOTE: Here, to simplify the analysis, we only consider the first run of each method\n",
    ")\n",
    "\n",
    "pb_set_dir = os.path.join(\n",
    "    \"..\",\n",
    "    \"data\",\n",
    "    \"posebusters_benchmark_set\",\n",
    ")\n",
    "assert os.path.exists(\n",
    "    pb_set_dir\n",
    "), \"Please download the PoseBusters Benchmark set from `https://zenodo.org/records/13858866` before proceeding.\"\n",
    "\n",
    "# Mappings\n",
    "method_mapping = {\n",
    "    \"diffdock\": \"DiffDock-L\",\n",
    "    \"diffdockv1\": \"-> w/o SCT\",\n",
    "    \"fabind\": \"FABind\",\n",
    "    \"dynamicbind\": \"DynamicBind\",\n",
    "    \"neuralplexer\": \"NeuralPLexer\",\n",
    "    \"rfaa\": \"RoseTTAFold-AA\",\n",
    "    \"chai-lab\": \"Chai-1\",\n",
    "    \"tulip\": \"TULIP\",\n",
    "    \"vina_diffdock\": \"DiffDock-L-Vina\",\n",
    "    \"vina_p2rank\": \"P2Rank-Vina\",\n",
    "    \"consensus_ensemble\": \"Ensemble (Con)\",\n",
    "}\n",
    "\n",
    "MAX_POSEBUSTERS_BENCHMARK_ANALYSIS_PROTEIN_SEQUENCE_LENGTH = 2000  # Only PoseBusters Benchmark targets with protein sequences below this threshold can be analyzed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ProteinSelect(Select):\n",
    "    \"\"\"A class to select only protein residues from a PDB file.\"\"\"\n",
    "\n",
    "    def accept_residue(self, residue: Any):\n",
    "        \"\"\"\n",
    "        Only accept residues that are part of a protein (e.g., standard amino acids).\n",
    "\n",
    "        :param residue: The residue to check.\n",
    "        :return: True if the residue is part of a protein, False otherwise.\n",
    "        \"\"\"\n",
    "        return residue.id[0] == \" \"  # NOTE: `HETATM` flag must be a blank for protein residues\n",
    "\n",
    "\n",
    "class LigandSelect(Select):\n",
    "    \"\"\"A class to select only ligand residues from a PDB file.\"\"\"\n",
    "\n",
    "    def accept_residue(self, residue: Any):\n",
    "        \"\"\"\n",
    "        Only accept residues that are part of a ligand.\n",
    "\n",
    "        :param residue: The residue to check.\n",
    "        :return: True if the residue is part of a ligand, False otherwise.\n",
    "        \"\"\"\n",
    "        return residue.id[0] != \" \"  # NOTE: `HETATM` flag must be a filled for ligand residues\n",
    "\n",
    "\n",
    "@beartype\n",
    "def create_temp_pdb_with_only_molecule_type_residues(\n",
    "    input_pdb_filepath: str,\n",
    "    molecule_type: Literal[\"protein\", \"ligand\"],\n",
    "    add_element_types: bool = False,\n",
    ") -> str:\n",
    "    \"\"\"\n",
    "    Create a temporary PDB file with only residues of a chosen molecule type.\n",
    "\n",
    "    :param input_pdb_filepath: The input PDB file path.\n",
    "    :param molecule_type: The molecule type to keep (either \"protein\" or \"ligand\").\n",
    "    :param add_element_types: Whether to add element types to the atoms.\n",
    "    :return: The temporary PDB file path.\n",
    "    \"\"\"\n",
    "    parser = PDBParser(QUIET=True)\n",
    "    structure = parser.get_structure(molecule_type, input_pdb_filepath)\n",
    "\n",
    "    io = PDBIO()\n",
    "    io.set_structure(structure)\n",
    "\n",
    "    # create a temporary PDB filepdb_name\n",
    "    temp_pdb_filepath = tempfile.NamedTemporaryFile(delete=False, suffix=\".pdb\")\n",
    "    io.save(\n",
    "        temp_pdb_filepath.name, ProteinSelect() if molecule_type == \"protein\" else LigandSelect()\n",
    "    )\n",
    "\n",
    "    if add_element_types:\n",
    "        with open(temp_pdb_filepath.name.replace(\".pdb\", \"_elem.pdb\"), \"w\") as f:\n",
    "            subprocess.run(  # nosec\n",
    "                f\"pdb_element {temp_pdb_filepath.name}\",\n",
    "                shell=True,\n",
    "                check=True,\n",
    "                stdout=f,\n",
    "            )\n",
    "        shutil.move(temp_pdb_filepath.name.replace(\".pdb\", \"_elem.pdb\"), temp_pdb_filepath.name)\n",
    "\n",
    "    return temp_pdb_filepath.name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute interaction fingerprints"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Analyze `PoseBusters Benchmark` set interactions as a baseline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"posebusters_benchmark_interaction_dataframes.h5\"):\n",
    "    pb_protein_ligand_filepath_pairs = []\n",
    "    for item in os.listdir(pb_set_dir):\n",
    "        ligand_item_path = os.path.join(pb_set_dir, item)\n",
    "        if os.path.isdir(ligand_item_path):\n",
    "            protein_filepath = os.path.join(ligand_item_path, f\"{item}_protein.pdb\")\n",
    "            ligand_filepath = os.path.join(ligand_item_path, f\"{item}_ligand.sdf\")\n",
    "            if os.path.exists(protein_filepath) and os.path.exists(ligand_filepath):\n",
    "                pb_protein_ligand_filepath_pairs.append((protein_filepath, ligand_filepath))\n",
    "\n",
    "    pc = (\n",
    "        PoseCheck()\n",
    "    )  # NOTE: despite what `PoseCheck` might say, `reduce` should be available in the `PoseBench` environment\n",
    "    pb_protein_ligand_interaction_dfs = []\n",
    "    for protein_filepath, ligand_filepath in tqdm(\n",
    "        pb_protein_ligand_filepath_pairs, desc=\"Processing PoseBusters Benchmark set\"\n",
    "    ):\n",
    "        temp_protein_filepath = create_temp_pdb_with_only_molecule_type_residues(\n",
    "            protein_filepath, molecule_type=\"protein\"\n",
    "        )\n",
    "        pc.load_protein_from_pdb(temp_protein_filepath)\n",
    "        pc.load_ligands_from_sdf(ligand_filepath)\n",
    "        pb_protein_ligand_interaction_dfs.append(pc.calculate_interactions())\n",
    "\n",
    "        # NOTE: we iteratively save the interaction dataframes to an HDF5 file\n",
    "        with pd.HDFStore(\"posebusters_benchmark_interaction_dataframes.h5\") as store:\n",
    "            for i, df in enumerate(pb_protein_ligand_interaction_dfs):\n",
    "                store.put(f\"df_{i}\", df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Analyze interactions of each method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate and cache PoseBusters Benchmark interaction statistics for each baseline method\n",
    "dataset = \"posebusters_benchmark\"\n",
    "for method in baseline_methods:\n",
    "    for repeat_index in range(1, max_num_repeats_per_method + 1):\n",
    "        method_title = method_mapping[method]\n",
    "\n",
    "        if not os.path.exists(f\"{method}_{dataset}_interaction_dataframes_{repeat_index}.h5\"):\n",
    "            v1_baseline = method == \"diffdockv1\"\n",
    "            base_method = (\n",
    "                method\n",
    "                if method == \"consensus_ensemble\"\n",
    "                else method.replace(\"v1\", \"\").split(\"_\")[0]\n",
    "            )\n",
    "            vina_binding_site_method = method.split(\"_\")[-1]\n",
    "\n",
    "            method_protein_dir = os.path.join(\n",
    "                resolve_method_protein_dir(\n",
    "                    base_method,\n",
    "                    dataset,\n",
    "                    repeat_index,\n",
    "                    pocket_only_baseline=False,\n",
    "                ),\n",
    "            )\n",
    "            method_ligand_dir = os.path.join(\n",
    "                \"..\",\n",
    "                resolve_method_ligand_dir(\n",
    "                    base_method,\n",
    "                    dataset,\n",
    "                    vina_binding_site_method,\n",
    "                    repeat_index,\n",
    "                    pocket_only_baseline=False,\n",
    "                    v1_baseline=v1_baseline,\n",
    "                ),\n",
    "            )\n",
    "\n",
    "            if not os.path.isdir(method_protein_dir):\n",
    "                method_protein_dir = os.path.join(\"..\", method_protein_dir)\n",
    "\n",
    "            if not os.path.isdir(method_ligand_dir):\n",
    "                # NOTE: this handles DynamicBind's output formats\n",
    "                ligand_dirs = [\n",
    "                    p\n",
    "                    for p in glob.glob(f\"{method_ligand_dir}_*_1/index0_idx_0\")\n",
    "                    if \"pocket_only\" not in p\n",
    "                ]\n",
    "            else:\n",
    "                ligand_dirs = [\n",
    "                    p\n",
    "                    for p in os.listdir(method_ligand_dir)\n",
    "                    if \"_relaxed\" not in p and \"bust_results\" not in p\n",
    "                ]\n",
    "\n",
    "            posebusters_protein_ligand_complex_filepaths = []\n",
    "            for item in ligand_dirs:\n",
    "                protein_item_path = os.path.join(method_protein_dir, f\"{item}*\")\n",
    "                ligand_item_path = os.path.join(method_ligand_dir, item)\n",
    "\n",
    "                complex_filepaths = []\n",
    "                if method == \"dynamicbind\":\n",
    "                    protein_item_path = glob.glob(os.path.join(item, \"rank1_receptor*.pdb\"))\n",
    "                    ligand_item_path = glob.glob(os.path.join(item, \"rank1_ligand*.sdf\"))\n",
    "                    if not protein_item_path or not ligand_item_path:\n",
    "                        continue\n",
    "                    complex_filepaths = [protein_item_path[0], ligand_item_path[0]]\n",
    "                elif os.path.isfile(ligand_item_path) and \"_relaxed\" not in item:\n",
    "                    # NOTE: this handles FABind's output formats\n",
    "                    complex_filepaths = glob.glob(\n",
    "                        os.path.join(\n",
    "                            os.path.dirname(protein_item_path),\n",
    "                            \"_\".join(item.split(\"_\")[:2]) + \".pdb\",\n",
    "                        )\n",
    "                    ) + glob.glob(ligand_item_path)\n",
    "                elif os.path.isdir(ligand_item_path) and \"_relaxed\" not in item:\n",
    "                    protein_pdb_filepath, ligand_sdf_filepath = None, None\n",
    "                    complex_filepaths = glob.glob(\n",
    "                        os.path.join(protein_item_path, \"*rank1*.pdb\")\n",
    "                    ) + glob.glob(os.path.join(ligand_item_path, \"*rank1.sdf\"))\n",
    "                    if not len(complex_filepaths) == 2:\n",
    "                        # NOTE: this handles DiffDock and TULIP's output formats\n",
    "                        complex_filepaths = glob.glob(f\"{protein_item_path}.pdb\") + glob.glob(\n",
    "                            os.path.join(ligand_item_path, \"rank1.sdf\")\n",
    "                        )\n",
    "                    if not len(complex_filepaths) == 2:\n",
    "                        # NOTE: this handles RFAA's output formats\n",
    "                        complex_filepaths = glob.glob(\n",
    "                            os.path.join(protein_item_path, \"*_protein.pdb\")\n",
    "                        ) + glob.glob(os.path.join(ligand_item_path, \"*_ligand.sdf\"))\n",
    "                    if not len(complex_filepaths) == 2:\n",
    "                        # NOTE: this handles Vina's output formats\n",
    "                        complex_filepaths = glob.glob(f\"{protein_item_path}.pdb\") + glob.glob(\n",
    "                            os.path.join(ligand_item_path, f\"{item}.sdf\")\n",
    "                        )\n",
    "\n",
    "                    if method == \"neuralplexer\":\n",
    "                        # NOTE: this handles NeuralPlexer's output formats\n",
    "                        complex_filepaths = [\n",
    "                            p\n",
    "                            for p in glob.glob(\n",
    "                                os.path.join(protein_item_path.removesuffix(\"*\"), \"*rank1_*.pdb\")\n",
    "                            )\n",
    "                            + glob.glob(\n",
    "                                os.path.join(ligand_item_path.removesuffix(\"*\"), \"*rank1_*.sdf\")\n",
    "                            )\n",
    "                            if \"relaxed\" not in p and \"aligned\" not in p\n",
    "                        ]\n",
    "                    elif method == \"chai-lab\":\n",
    "                        # NOTE: this handles Chai-1's output formats\n",
    "                        complex_filepaths = [\n",
    "                            p\n",
    "                            for p in glob.glob(\n",
    "                                os.path.join(\n",
    "                                    protein_item_path.removesuffix(\"*\"),\n",
    "                                    \"pred.model_idx_0_protein.pdb\",\n",
    "                                )\n",
    "                            )\n",
    "                            + glob.glob(\n",
    "                                os.path.join(\n",
    "                                    ligand_item_path.removesuffix(\"*\"),\n",
    "                                    \"pred.model_idx_0_ligand.sdf\",\n",
    "                                )\n",
    "                            )\n",
    "                        ]\n",
    "                    elif method == \"consensus_ensemble\":\n",
    "                        # NOTE: this handles the Consensus Ensemble's output formats\n",
    "                        complex_filepaths = glob.glob(\n",
    "                            os.path.join(protein_item_path.removesuffix(\"*\"), \"*.pdb\")\n",
    "                        ) + glob.glob(os.path.join(ligand_item_path.removesuffix(\"*\"), \"*.sdf\"))\n",
    "\n",
    "                if not len(complex_filepaths) == 2:\n",
    "                    continue\n",
    "\n",
    "                for file in complex_filepaths:\n",
    "                    if file.endswith(\".pdb\"):\n",
    "                        protein_pdb_filepath = file\n",
    "                    elif file.endswith(\".sdf\"):\n",
    "                        ligand_sdf_filepath = file\n",
    "                if protein_pdb_filepath is not None and ligand_sdf_filepath is not None:\n",
    "                    posebusters_protein_ligand_complex_filepaths.append(\n",
    "                        (protein_pdb_filepath, ligand_sdf_filepath)\n",
    "                    )\n",
    "                else:\n",
    "                    print(\n",
    "                        f\"Warning: Could not find `rank1` protein-ligand complex files for {item}. Skipping...\"\n",
    "                    )\n",
    "\n",
    "            pc = (\n",
    "                PoseCheck()\n",
    "            )  # NOTE: despite what `PoseCheck` might say, `reduce` should be available in the `PoseBench` environment\n",
    "            posebusters_protein_ligand_interaction_dfs = []\n",
    "            for protein_ligand_complex_filepath in tqdm(\n",
    "                posebusters_protein_ligand_complex_filepaths,\n",
    "                desc=f\"Processing interactions for {method_title}\",\n",
    "            ):\n",
    "                try:\n",
    "                    protein_filepath, ligand_filepath = protein_ligand_complex_filepath\n",
    "                    num_residues_in_target_protein = count_num_residues_in_pdb_file(\n",
    "                        protein_filepath\n",
    "                    )\n",
    "                    if (\n",
    "                        num_residues_in_target_protein\n",
    "                        > MAX_POSEBUSTERS_BENCHMARK_ANALYSIS_PROTEIN_SEQUENCE_LENGTH\n",
    "                    ):\n",
    "                        print(\n",
    "                            f\"{method_title} target {protein_ligand_complex_filepath} has too many protein residues ({num_residues_in_target_protein} > {MAX_POSEBUSTERS_BENCHMARK_ANALYSIS_PROTEIN_SEQUENCE_LENGTH}) for `MDAnalysis` to fit into CPU memory. Skipping...\"\n",
    "                        )\n",
    "                        continue\n",
    "                    ligand_mol = Chem.MolFromMolFile(ligand_filepath)\n",
    "                    pc.load_protein_from_pdb(protein_filepath)\n",
    "                    pc.load_ligands_from_mols(\n",
    "                        Chem.GetMolFrags(ligand_mol, asMols=True, sanitizeFrags=False)\n",
    "                    )\n",
    "                    posebusters_protein_ligand_interaction_dfs.append(pc.calculate_interactions())\n",
    "                except Exception as e:\n",
    "                    print(\n",
    "                        f\"Error processing {method_title} target {protein_ligand_complex_filepath} due to: {e}. Skipping...\"\n",
    "                    )\n",
    "                    continue\n",
    "\n",
    "                # NOTE: we iteratively save the interaction dataframes to an HDF5 file\n",
    "                with pd.HDFStore(\n",
    "                    f\"{method}_posebusters_benchmark_interaction_dataframes_{repeat_index}.h5\"\n",
    "                ) as store:\n",
    "                    for i, df in enumerate(posebusters_protein_ligand_interaction_dfs):\n",
    "                        store.put(f\"df_{i}\", df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot interaction statistics for each method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "\n",
    "\n",
    "# define a function to process each method\n",
    "def process_method(file_path, category):\n",
    "    interactions = []\n",
    "    with pd.HDFStore(file_path) as store:\n",
    "        for key in store.keys():\n",
    "            for row_index in range(len(store[key])):\n",
    "                interaction_types = [\n",
    "                    interaction[2] for interaction in store[key].iloc[row_index].keys().tolist()\n",
    "                ]\n",
    "                num_hb_acceptors = interaction_types.count(\"HBAcceptor\")\n",
    "                num_hb_donors = interaction_types.count(\"HBDonor\")\n",
    "                num_vdw_contacts = interaction_types.count(\"VdWContact\")\n",
    "                num_hydrophobic = interaction_types.count(\"Hydrophobic\")\n",
    "                interactions.append(\n",
    "                    {\n",
    "                        \"Hydrogen Bond Acceptors\": num_hb_acceptors,\n",
    "                        \"Hydrogen Bond Donors\": num_hb_donors,\n",
    "                        \"Van der Waals Contacts\": num_vdw_contacts,\n",
    "                        \"Hydrophobic Interactions\": num_hydrophobic,\n",
    "                    }\n",
    "                )\n",
    "    df_rows = []\n",
    "    for interaction in interactions:\n",
    "        for interaction_type, num_interactions in interaction.items():\n",
    "            df_rows.append(\n",
    "                {\n",
    "                    \"Category\": category,\n",
    "                    \"InteractionType\": interaction_type,\n",
    "                    \"NumInteractions\": num_interactions,\n",
    "                }\n",
    "            )\n",
    "    return pd.DataFrame(df_rows)\n",
    "\n",
    "\n",
    "# load data from files\n",
    "for method in baseline_methods:\n",
    "    for repeat_index in range(1, max_num_repeats_per_method + 1):\n",
    "        method_title = method_mapping[method]\n",
    "        file_path = f\"{method}_posebusters_benchmark_interaction_dataframes_{repeat_index}.h5\"\n",
    "        if os.path.exists(file_path):\n",
    "            dfs.append(process_method(file_path, method_title))\n",
    "\n",
    "if os.path.exists(\"posebusters_benchmark_interaction_dataframes.h5\"):\n",
    "    dfs.append(process_method(\"posebusters_benchmark_interaction_dataframes.h5\", \"Reference\"))\n",
    "\n",
    "# combine statistics\n",
    "assert len(dfs) > 0, \"No interaction dataframes found.\"\n",
    "df = pd.concat(dfs)\n",
    "\n",
    "# define font properties\n",
    "plt.rcParams[\"font.size\"] = 14\n",
    "plt.rcParams[\"axes.labelsize\"] = 16\n",
    "\n",
    "# plot statistics\n",
    "fig, axes = plt.subplots(2, 2, figsize=(34, 14), sharey=False)\n",
    "\n",
    "interaction_types = [\n",
    "    \"Hydrogen Bond Acceptors\",\n",
    "    \"Hydrogen Bond Donors\",\n",
    "    \"Van der Waals Contacts\",\n",
    "    \"Hydrophobic Interactions\",\n",
    "]\n",
    "plot_types = [\"box\", \"box\", \"violin\", \"violin\"]\n",
    "\n",
    "for ax, interaction, plot_type in zip(axes.flatten(), interaction_types, plot_types):\n",
    "    data = df[df[\"InteractionType\"] == interaction]\n",
    "\n",
    "    if plot_type == \"box\":\n",
    "        sns.boxplot(data=data, x=\"Category\", y=\"NumInteractions\", ax=ax, showfliers=True)\n",
    "        sns.stripplot(\n",
    "            data=data,\n",
    "            x=\"Category\",\n",
    "            y=\"NumInteractions\",\n",
    "            ax=ax,\n",
    "            color=\"black\",\n",
    "            alpha=0.3,\n",
    "            jitter=True,\n",
    "        )\n",
    "    elif plot_type == \"violin\":\n",
    "        sns.violinplot(data=data, x=\"Category\", y=\"NumInteractions\", ax=ax)\n",
    "        sns.stripplot(\n",
    "            data=data,\n",
    "            x=\"Category\",\n",
    "            y=\"NumInteractions\",\n",
    "            ax=ax,\n",
    "            color=\"black\",\n",
    "            alpha=0.3,\n",
    "            jitter=True,\n",
    "        )\n",
    "\n",
    "    ax.set_title(interaction)\n",
    "    ax.set_ylabel(\"No. Interactions\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.grid(True)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"posebusters_benchmark_method_interaction_analysis.png\", dpi=300)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PoseBench",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
